File: tests/fixtures/open_redirect/typescript/safe_host_allowlist_redirect.ts
// Safe: req.query.next routed through `new URL(...).hostname === ALLOWED`
// host-allowlist gate before reaching res.redirect.  Recognised by
// PredicateKind::HostAllowlistValidated which clears Cap::OPEN_REDIRECT
// on the validated branch.
const ALLOWED_HOST: string = "trusted.example.com";

export function handler(req: any, res: any): void {
    const target: string = req.query.next;
    if (new URL(target).hostname === ALLOWED_HOST) {
        res.redirect(target);
        return;
    }
    res.redirect("/");
}

File: tests/fixtures/open_redirect/python/safe_host_allowlist_redirect.py
# Safe: query arg routed through `urlparse(...).netloc == ALLOWED`
# host-allowlist gate before passing to flask.redirect.  Recognised by
# PredicateKind::HostAllowlistValidated which clears Cap::OPEN_REDIRECT
# on the validated branch.
from flask import request, redirect
from urllib.parse import urlparse

ALLOWED_HOST = "trusted.example.com"


def handler():
    target = request.args.get("next")
    if urlparse(target).netloc == ALLOWED_HOST:
        return redirect(target)
    return redirect("/")
